### DISCUSSION TITLE
vscode TODO extension question

### DISCUSSION THREAD

**Discussion by SamiranRai** [microsoft/vscode-discussions#1150] [Extension Development QnA] [answered]
I am making a vscode TODO extension with some cool features. I am using "vscode.Range" and decoration for the highlighting process. everything is work fine! but the main problem is when I highlight certain text like // NAME: -it highlight the N index to : Index and let's assume the N index is 3 and semicolon index is 7. so if my cursor is present in 2 or 8 index the highlight is continue and if I console the range property its changing but the highlight is not properly executed according to the change! I want if I press space the highlight also be spaced from both side left and right! please provide me solution?

**Comment by theReynald** [comment, marked as answer]
Hi @samiranrai.

I updated the title on this to be more concise rather then having multiple sentences.

It sounds like you're encountering an issue with updating the visual highlight in response to cursor position changes in your Visual Studio Code extension. Here's a step-by-step solution to help you adjust the highlight dynamically when the cursor position changes, such as when you press space near the highlighted text.

1. **Update Decoration on Cursor Move:**
   You need to handle the cursor movement and update the decoration accordingly. VS Code provides events for cursor position changes which you can use to trigger updates to your decorations.

2. **Calculate New Range:**
   When the cursor moves, you'll need to recalculate the range of your decoration. If the cursor is next to the highlight and a space is inserted, you should adjust the start and/or end of the range accordingly.

3. **Apply New Decoration:**
   After recalculating the range, you should apply a new decoration to reflect the updated range. It's essential to remove the old decoration first to avoid overlapping or stale highlights.

Here’s a basic implementation in TypeScript that you might find useful:

```typescript
import * as vscode from 'vscode';

export function activate(context: vscode.ExtensionContext) {
    let activeEditor = vscode.window.activeTextEditor;
    let decorationType = vscode.window.createTextEditorDecorationType({
        backgroundColor: 'rgba(220, 220, 170, 0.5)' // example highlight color
    });

    function updateDecorations() {
        if (!activeEditor) {
            return;
        }

        const text = activeEditor.document.getText();
        const regex = /\/\/ NAME:/g; // regex to match your pattern
        const decorations: vscode.DecorationOptions[] = [];

        let match;
        while (match = regex.exec(text)) {
            const startPos = activeEditor.document.positionAt(match.index);
            const endPos = activeEditor.document.positionAt(match.index + match[0].length);
            const decoration = { range: new vscode.Range(startPos, endPos) };
            decorations.push(decoration);
        }

        activeEditor.setDecorations(decorationType, decorations);
    }

    if (activeEditor) {
        updateDecorations();
    }

    vscode.window.onDidChangeActiveTextEditor(editor => {
        activeEditor = editor;
        if (editor) {
            updateDecorations();
        }
    }, null, context.subscriptions);

    vscode.workspace.onDidChangeTextDocument(event => {
        if (activeEditor && event.document === activeEditor.document) {
            updateDecorations();
        }
    }, null, context.subscriptions);

    vscode.window.onDidChangeTextEditorSelection(event => {
        if (activeEditor && event.textEditor === activeEditor) {
            updateDecorations(); // call updateDecorations to recalculate the highlight based on new cursor position
        }
    }, null, context.subscriptions);
}

export function deactivate() {}
```

### Key Points:
- **`onDidChangeTextEditorSelection`:** This event is triggered when the cursor moves, which is essential for recalculating the decorations when your cursor moves next to the highlighted section.
- **`updateDecorations`:** This function recalculates and updates the decorations based on the current state of the text in the editor.

This script should give you a good foundation to modify and adapt based on your specific requirements, such as adjusting the ranges when spaces are inserted next to the highlighted text.

Is this along the lines of what you wanted to know? if not, please rephrase your questions.

**Comment by abrilurena** [comment]
Hi, @SamiranRai! Do you still need help with this? thanks!
